doc/developer/design/aphyr-txn-notes.md

```diff
@@ -0,0 +1,115 @@
+We have been chatting with Kyle Kingsbury about the design of a transaction
+protocol in a world where.
+
+# Calvin-inspired
+
+## Background
+
+Calvin is a distributed database transaction protocol that relies on
+deterministic execution to avoid coordination as much as possible. In the
+prototypical Calvin implementation global coordination occurs at most once per
+epoch (e.g. every 10 ms), where each client-request-receiving service (known as
+the sequencer) shares the order of execution it determined with other all other
+the services that depend on that order (known as scheulders). Because the
+execution is deterministic, ensuring each scheduler receives the same set of
+inputs means they will eventually converge to the same final states.
+
+## Materialize design
+
+Materialize propagates all write operations to a durable WAL. (Note that unlike
+Calvin, we only propagate writes to the WAL and not the entire transactional
+input; we are not running full stack of Materialize services vertically.)
+
+Storage nodes are then responsible for taking updates to the WAL and attempting
+to apply them to their local state--note that this, in essence, pushes down
+transaction logic to the storage layer. Storage nodes then need some means to
+communicate with one another to either all accept or all reject the update. In
+the case of rejection, the storage node needs some means to discard the
+provisional writes.
+
+Note that in this design, some external source is responsible for determining
+the order in which writes get appended to the WAL, i.e. the WAL itself is truly
+just a log that offers to real invariants itself.
+
+This design is inspired by Calvin in as much as the storage nodes get to behave
+akin t
+
+### Example architecture
+
+```
+  client
+  |    ^
+ txnA  |
+  |   res
+  v    |                 storageA    storageB
+[ADAPTER]-wA->[WAL]<-get---?
+^              ^-----------get-----------?
+|
+|              if a?.is_ok() && b?.is_ok() {
+|                  return OK to client
+|              } else {
+|                  rollback(A);
+|                  rollback(B);
+|                  return ABORT to client
+|______________}
+```
+
+### Questions
+
+- How do writes invalidate outstanding reads in e.g. `UPDATE`? I realize this
+  design says that sequencing has to happen elsewhere, but just curious where we
+  envision that potentially being.
+- How do e.g. column constraints get handled by the txn layer? Seems like this
+  moves "business logic" into storage, which I don't think will make the storage
+  people happy.
+- How do we rollback if we're already in the storage layer?
+
+# CAS
+
+## Background
+
+This is the design Kyle has been moving forward with fleshing out, which relies
+on a the WAL having a timestamp only ever written to atomically.
+
+## Materialize design
+
+Each writable TVC will have a WAL that acts as a buffer for writes before being
```

Comment:
Oop! I think it's that there is still a single WAL, not one per TVC. The WAL appends all writes an a transaction atomically, then storage nodes are responsible for consuming it. The difference between the CAS and non CAS versions is that in the CAS version, if a txn exists in the WAL it committed. In the non CAS version consumers of the WAL must do something to decide if the txn committed (I think this was to have in the txn data a "previous WAL operation timestamp must be X" value which can checked in the WAL, or maybe some additional coordination between all storage nodes (I'm less sure about this)).

Reply:
This was totally right and an oversight on my part.